### PR TITLE
971 Flatten block structure

### PR DESCRIPTION
**Pull request by mfaulk**
Changes the block structure from a list of redacted transactions to a list of key images and a list of outputs.

Still a few TODOs and disabled tests to fix.

### CODE REVIEW

File: consensus/api/proto/blockchain.proto [hunk 1]
```diff
@@ -51,8 +59,8 @@ message S3Block {
     // Block
     Block block = 1;
 
-    // Transactions in this block
-    repeated external.RedactedTx transactions = 2;
+    // Contents of the block.
+    BlockContents block_contents = 2;
```
Comment (mfaulk) [inline consensus/api/proto/blockchain.proto:63]: Conceptually, this is the main change.

File: consensus/enclave/impl/src/lib.rs [hunk 2]
```diff
@@ -462,44 +460,47 @@ impl ConsensusEnclave for SgxConsensusEnclave {
...
-            redacted_transactions.push(tx.redact());
+        let mut outputs: Vec<TxOut> = Vec::new();
+        let mut key_images: Vec<KeyImage> = Vec::new();
+        for tx in &transactions {
+            outputs.extend(tx.prefix.outputs.iter().cloned());
+            key_images.extend(tx.key_images().iter().cloned());
         }
+        outputs.push(fee_output);
+
+        // TODO: sort outputs and key images.
```
Comment (mfaulk) [inline consensus/enclave/impl/src/lib.rs:473]: It might be best to do this sorting in a separate PR since reordering the outputs and key images will probably break a lot of unit tests.

File: consensus/enclave/impl/src/lib.rs [hunk 3]
```diff
@@ -802,47 +787,49 @@ mod tests {
...
-            .unwrap();
-        assert_eq!(value, total_fee);
-
-        // Each of the input transactions should be redacted.
-        for (i, tx) in input_transactions.into_iter().enumerate() {
-            let expected = tx.redact();
-
-            assert_eq!(expected, redacted_transactions[i + 1]);
-        }
+        // TODO: update the following tests.
```
Comment (mfaulk) [inline consensus/enclave/impl/src/lib.rs:790]: Will do

File: transaction/core/src/lib.rs [hunk 4]
```diff
@@ -19,7 +19,7 @@ use keys::{RistrettoPrivate, RistrettoPublic};
 pub mod account_keys;
 pub mod amount;
 pub mod blake2b_256;
-mod block;
+mod blockchain;
```
Comment (mfaulk) [inline transaction/core/src/lib.rs:22]: This PR splits block.rs into a few separate files. I find small files easier to navigate, and easier to get a sense of what has test coverage.

File: transaction/core/src/blockchain/block_contents.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
+use crate::{blake2b_256::Blake2b256, ring_signature::KeyImage, tx::TxOut, ConvertError};
+use alloc::vec::Vec;
+use core::convert::TryFrom;
+use digestible::{Digest, Digestible};
+use generic_array::{typenum::Unsigned, GenericArray};
+use serde::{Deserialize, Serialize};
+
+/// The contents of a Block.
+#[derive(Clone, PartialEq, Eq, Debug, Digestible, Serialize, Deserialize)]
+pub struct BlockContents {
```
Comment (mfaulk) [inline transaction/core/src/blockchain/block_contents.rs:10]: This type generally replaces Vec<RedactedTx> wherever we are working with the contents of a block.

File: consensus/api/src/conversions.rs [hunk 6]
```diff
@@ -870,6 +812,99 @@ impl TryInto<TransactionValidationError> for ProposeTxResult {
...
+        let key_images: Vec<external::KeyImage> = source
+            .key_images
+            .iter()
+            .map(external::KeyImage::from)
+            .collect();
+
+        let outputs = source.outputs.iter().map(external::TxOut::from).collect();
+
+        block_contents.set_key_images(RepeatedField::from_vec(key_images));
+        block_contents.set_outputs(outputs);
```
Comment (eranrund) [inline consensus/api/src/conversions.rs:865]: Curious why one call needs `RepeatedField` and the other doesn't?

File: ledger/distribution/src/main.rs [hunk 7]
```diff
@@ -145,20 +145,23 @@ impl BlockHandler for S3BlockWriter {
...
         log::info!(self.logger, "S3: Handling block {}", block.index);
 
         let bc_block = blockchain::Block::from(block);
-        let bc_transactions = transactions
-            .iter()
-            .map(external::RedactedTx::from)
-            .collect();
+        let bc_block_contents = blockchain::BlockContents::from(block_contents);
+
+        // let bc_transactions = transactions
```
Comment (eranrund) [inline ledger/distribution/src/main.rs:156]: Leftovers

File: ledger/distribution/src/main.rs [hunk 8]
```diff
@@ -200,20 +203,22 @@ impl BlockHandler for LocalBlockWriter {
...
     ) {
         log::info!(self.logger, "S3: Handling block {}", block.index);
 
         let bc_block = blockchain::Block::from(block);
-        let bc_transactions = transactions
-            .iter()
-            .map(external::RedactedTx::from)
-            .collect();
+        let bc_block_contents = blockchain::BlockContents::from(block_contents);
+        // let bc_transactions = transactions
```
Comment (eranrund) [inline ledger/distribution/src/main.rs:213]: Leftovers

File: ledger/sync/src/ledger_sync_service.rs [hunk 9]
```diff
@@ -280,33 +279,33 @@ impl<L: Ledger, BC: BlockchainConnection + 'static, TF: TransactionsFetcher + 's
...
             {
                 let _timer = counters::APPEND_BLOCK_TIME.start_timer();
-                self.ledger.append_block(block, transactions, None)?;
+                self.ledger.append_block(block, contents, None)?;
             }
 
             // FIXME: MC-365 Move ledger counters into ledger_db
             counters::BLOCKS_WRITTEN_COUNT.inc();
-            counters::TX_WRITTEN_COUNT.inc_by(transactions.len() as i64);
+            // counters::TX_WRITTEN_COUNT.inc_by(transactions.len() as i64);
```
Comment (eranrund) [inline ledger/sync/src/ledger_sync_service.rs:306]: Looks like we don't have TXO_WRITTEN_COUNT so this metric should probably be adapted to that. and the TX_IN_LEDGER one below (and wherever else it appears) removed.

File: ledger/sync/src/reqwest_transactions_fetcher.rs [hunk 10]
```diff
@@ -117,30 +117,40 @@ impl ReqwestTransactionsFetcher {
...
                     url.to_string(),
-                    format!("tx conversion failed: {:?}", err),
+                    format!("Block contents conversion failed: {:?}", err),
                 )
             })?;
-            redacted_transactions.push(redacted_tx);
-        }
 
-        let signature = bc_block
+        // let mut redacted_transactions = Vec::new();
```
Comment (eranrund) [inline ledger/sync/src/reqwest_transactions_fetcher.rs:142]: Leftovers

File: ledger/sync/src/reqwest_transactions_fetcher.rs [hunk 11]
```diff
@@ -164,34 +174,34 @@ impl ReqwestTransactionsFetcher {
 
         let s3_block_data = S3BlockData {
```
Comment (eranrund) [inline ledger/sync/src/reqwest_transactions_fetcher.rs:175]: There should probably be a check here that verifies that block.contents_hash == block_contents.hash()

File: transaction/core/src/blockchain/block_contents.rs (hunk 5 above)
Comment (eranrund) [inline transaction/core/src/blockchain/block_contents.rs:10]: Can this derive `Message`? That would simplify things down the line (I think it should just work since TxOut and KeyImage derive it?)

**Comment by mfaulk** [comment]
Thanks for the reviews, @eranrund.

Regarding the BlockContents database: it does store a redundant copy of all ledger data, but it replaces the "transactions_by_block" database which also stored a redundant copy of all ledger data. I'm in favor of not storing duplicate data, but would prefer to do that in a separate PR.

File: consensus/api/src/conversions.rs (hunk 6 above)
Comment (mfaulk) [reply consensus/api/src/conversions.rs:865]: Good catch

File: transaction/std/src/lib.rs [hunk 12]
```diff
@@ -1,12 +1,10 @@
 // Copyright (c) 2018-2020 MobileCoin Inc.
 
-mod block_builder;
 mod error;
 pub mod identity;
 mod input_credentials;
 mod transaction_builder;
 
-pub use block_builder::BlockBuilder;
```
Comment (garbageslam) [inline transaction/std/src/lib.rs:9]: i'm glad to see this guy go away, it seems like there wasn't really any uptake here

File: util/generate-sample-ledger/src/lib.rs [hunk 13]
```diff
@@ -44,85 +44,60 @@ pub fn bootstrap_ledger(
...
     let mut previous_block: Option<Block> = None;
 
+    let mut rng: FixedRng = SeedableRng::from_seed([33u8; 32]);
+
     for block_index in 0..num_blocks as u64 {
         println!("Creating block {} of {}.", block_index, num_blocks);
 
-        // Transactions in this block.
-        let minting_transactions: Vec<RedactedTx> = recipients
-            .par_iter()
```
Comment (garbageslam) [inline util/generate-sample-ledger/src/lib.rs:55]: do you know what the `par_iter` thing here was? i think it was to try to make bootstrap faster because it can be really slow when the numbers are big i think

Comment (mfaulk) [reply util/generate-sample-ledger/src/lib.rs:55]: I added that at one point to speed things up, but I think the real problem was that I had forgotten to compile bootstrap in release mode.

File: ledger/db/src/lib.rs [hunk 14]
```diff
@@ -369,67 +341,57 @@ impl LedgerDB {
...
         }
 
-        // Check if block is being appended at the correct place
+        // A block must have outputs.
+        if block_contents.outputs.is_empty() {
+            // TODO: better error type.
+            return Err(Error::InvalidBlock);
+        }
+
+        // TODO: enable this.
```
Comment (mfaulk) [inline ledger/db/src/lib.rs:361]: I'd like to add this new check. It breaks several unit tests, so I'd like to deal with it as a follow-on PR,
https://mobilecoin.atlassian.net/browse/MC-1335

File: ledger/db/src/lib.rs [hunk 15]
```diff
@@ -779,360 +664,221 @@ mod ledger_db_test {
...
-
-        // The ledger should contain `key_image_b`.
-        assert!(ledger_db.contains_key_image(&key_image_b).unwrap());
-
-        let key_image_c: KeyImage = KeyImage::from(3);
-
-        // The ledger should not contain `key_image_c`.
-        assert_eq!(false, ledger_db.contains_key_image(&key_image_c).unwrap());
-    }
+    // #[test]
```
Comment (mfaulk) [inline ledger/db/src/lib.rs:716]: Will re-enable this.

File: consensus/enclave/impl/src/lib.rs (hunk 2 above)
Comment (mfaulk) [reply consensus/enclave/impl/src/lib.rs:473]: https://mobilecoin.atlassian.net/browse/MC-1337
